cpp/tests/community/k_truss_test.cpp

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governin_from_mtxg permissions and
+ * limitations under the License.
+ */
+
+#include "utilities/check_utilities.hpp"
+#include "utilities/conversion_utilities.hpp"
+
+#include <cugraph/algorithms.hpp>
+#include <cugraph/graph.hpp>
+#include <cugraph/graph_functions.hpp>
+#include <cugraph/graph_view.hpp>
+#include <cugraph/utilities/high_res_timer.hpp>
+
+#include <raft/core/handle.hpp>
+#include <raft/util/cudart_utils.hpp>
+
+#include <gtest/gtest.h>
+#include <utilities/base_fixture.hpp>
+#include <utilities/test_graphs.hpp>
+
+#include <algorithm>
+#include <iterator>
+#include <limits>
+#include <numeric>
+#include <random>
+#include <set>
+#include <vector>
+
+struct KTruss_Usecase {
+  int32_t k_{3};
+  bool test_weighted_{false};
+  bool check_correctness_{true};
+};
+
+template <typename input_usecase_t>
+class Tests_KTruss : public ::testing::TestWithParam<std::tuple<KTruss_Usecase, input_usecase_t>> {
+ public:
+  Tests_KTruss() {}
+
+  static void SetUpTestCase() {}
+  static void TearDownTestCase() {}
+
+  virtual void SetUp() {}
+  virtual void TearDown() {}
+
+  // FIXME: There is an utility equivalent functor not
+  // supporting host vectors.
+  template <typename type_t>
+  struct device_nearly_equal {
+    const type_t threshold_ratio;
+    const type_t threshold_magnitude;
+
+    bool operator()(type_t lhs, type_t rhs) const
+    {
+      return std::abs(lhs - rhs) <
+             std::max(std::max(lhs, rhs) * threshold_ratio, threshold_magnitude);
+    }
+  };
+
+  template <typename vertex_t, typename edge_t, typename weight_t>
+  std::tuple<std::vector<vertex_t>, std::vector<vertex_t>, std::optional<std::vector<weight_t>>>
+  k_truss_reference(
+    raft::handle_t const& handle,
+    cugraph::graph_view_t<vertex_t, edge_t, false, false> const& graph_view,
+    std::optional<cugraph::edge_property_view_t<edge_t, weight_t const*>> edge_weight_view,
+    edge_t k)
+  {
+    auto [h_offsets, h_indices, h_values] = cugraph::test::graph_to_host_csr(
+      handle,
+      graph_view,
+      edge_weight_view,
+      std::optional<raft::device_span<vertex_t const>>(std::nullopt));
```

Comment:
Our convention is to call this outside the reference implementation and keep the reference implementation completely independent of any device functions or anything specific to cugraph.